### DISCUSSION TITLE
Using jsPsych.data.addProperties() on cognition.run

### DISCUSSION THREAD

**Discussion by amunn** [jspsych/jsPsych#3000] [Q&A] [answered]
We collect demographic data using the `survey-html-form` plugin. When running locally or on our JATOS server, the following function, added to the `on_finish:` parameter of `initJsPsych()` correctly adds the data to each trial. This doesn't work on cognition.run.  There is no error, but the data simply doesn't get written.

What do I need to do to get this to work there?

```
function add_demo_data()
		{
			demo_data = jsPsych.data.get().filter({trial_type: 'survey-html-form'}).values()[0].response;
			jsPsych.data.addProperties({
				gender: demo_data.gender,
				parent_1_gender: demo_data.parent_1_gender,
				parent_2_gender: demo_data.parent_2_gender,
				parent_1_ed: demo_data.parent_1_ed,
				parent_2_ed: demo_data.parent_2_ed,
				date_of_birth: demo_data.DOB,
				zipcode: demo_data.zipcode,
				first_lang: demo_data.first_lang
	})}; 
```

**Comment by ChristopheBossens** [comment]
I have always been curious as to how exactly cognition.run approaches the saving of data, as there is no additional code that you need to add in order for that to work. So without knowing that I am not really sure what the answer to this issue would be.

Where in the timeline of your experiment do you have the demographics? One thing I noticed is that cognition.run saves on a trial by trial basis, so it would make sense that you do not see the data added at least for trials before the demographics form.

One thing I would try is create an on_finish() function for a specific trial and add some dummy data just to see if that is handled (see [here](https://www.jspsych.org/7.3/overview/plugins/#the-on_finish-parameter)).

**Reply by amunn** [reply, marked as answer]
@ChristopheBossens  Thanks for this. Yes, it seems that it is saving data trial by trial, which would account for the data not being written using the experiment level `on_finish` parameter. Experimenting a bit, it seems that if I simply add my function to the `on_finish` parameter of the demographic trial itself, it correctly adds the date to each subsequent trial. The key phrase in the  documentation of `jsPsych.data.addProperties`  (which I hadn't noticed before) is (emphasis mine):

> This method appends a set of properties to every trial in the data object, including trials that have already occurred and **trials that have yet to occur**. 

This means that my demographic trial can simply be:

```
var demographic_trial = {
			type: jsPsychSurveyHtmlForm,
			preamble: "<p>Before we begin, we'd like to know a bit about you.</p>" +
								"<p>Please answer the following questions:</p>",
			html: questions,
			on_finish: function() {add_demo_data()}
		};
```

And the data will get added to all subsequent trials as well, which is exactly what I need.  Since this method will work equally well on JATOS, I can also keep my code consistent across collection platforms, which is helpful for my sanity. :)
